compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -968,30 +968,31 @@ object PatternMatcher {
     /** If match is switch annotated, check that it translates to a switch
      *  with at least as many cases as the original match.
      */
-    private def checkSwitch(original: Match, result: Tree) = original.selector match {
+    private def checkSwitch(original: Match, result: Tree) = original.selector match
       case Typed(_, tpt) if tpt.tpe.hasAnnotation(defn.SwitchAnnot) =>
-        val resultCases = result match {
+        val resultCases = result match
           case Match(_, cases) => cases
           case Block(_, Match(_, cases)) => cases
+          case Block((_: ValDef) :: Block(_, Match(_, cases)) :: Nil, _) => cases
           case _ => Nil
-        }
-        def typesInPattern(pat: Tree): List[Type] = pat match {
+        val caseThreshold =
+          if ValueClasses.isDerivedValueClass(tpt.tpe.typeSymbol) then 1
+          else MinSwitchCases
```

Comment:
Why do we have this branch? Shouldn't it always be 1?

Reply:
The general idea is to limit the warnings to when `numTypes(original.cases) >= MinSwitchCases`

https://github.com/lampepfl/dotty/blob/609d8d6bdbe9be33080967cdac478a6957237356/compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala#L59-L60

However value classes are special cased as an attempt at "best-effort" support. Consider:

```scala
case class IntAnyVal(x: Int) extends AnyVal

final val Ten = IntAnyVal(10)

def test(x: IntAnyVal) = (x: @switch) match {
  case IntAnyVal(1) => 0
  case Ten           => 1
  case IntAnyVal(100) => 2
  case IntAnyVal(1000) => 3
  case IntAnyVal(10000) => 4
}
```

which results in

| Item | Value|Notes
|------|-------|---
`numTypes(original.cases)` | `2`|`{ IntAnyVal, Ten.type }`
`numTypes(resultCases)`|`0`
emitted bytecode|if-then

We want to warn here since the `switch` was not emitted despite a sufficient number of cases, thus we use a lower `caseThreshold` for value classes.